File: neo/frontend/lib/api/knlg-base/knowledge.ts
/**
 * Knowledge Base API client.
 */

import {
	knlgBasePath,
	knlgDelete,
	knlgGet,
	knlgPost,
	knlgPut,
	type KnowledgeCard,
	type KnowledgeCardCreate,
	type KnowledgeCardListResponse,
	type KnowledgeCardUpdate,
	type SourceRef,
} from "./_base";

const path = (ws: string, suffix = "") =>
	knlgBasePath(ws, `/knowledge${suffix}`);

export async function listKnowledgeCards(
	workspaceCode: string,
	params?: {
		page?: number;
		page_size?: number;
		domain?: string;
		type?: string;
		status?: string;
		validation_status?: string;
		keyword?: string;
	},
): Promise<KnowledgeCardListResponse> {
	const search = new URLSearchParams();
	if (params) {
		Object.entries(params).forEach(([k, v]) => {
			if (v !== undefined && v !== null) search.append(k, String(v));
		});
	}
	const qs = search.toString();
	return knlgGet<KnowledgeCardListResponse>(
		path(workspaceCode, qs ? `?${qs}` : ""),
	);
}

export async function getKnowledgeCard(
	workspaceCode: string,
	id: number,
): Promise<KnowledgeCard> {
	return knlgGet<KnowledgeCard>(path(workspaceCode, `/cards/${id}`));
}

export async function createKnowledgeCard(
	workspaceCode: string,
	data: KnowledgeCardCreate,
): Promise<KnowledgeCard> {
	return knlgPost<KnowledgeCard>(path(workspaceCode), data);
}

export async function updateKnowledgeCard(
	workspaceCode: string,
	id: number,
	data: KnowledgeCardUpdate,
): Promise<KnowledgeCard> {
	return knlgPut<KnowledgeCard>(path(workspaceCode, `/cards/${id}`), data);
}

export async function deleteKnowledgeCard(
	workspaceCode: string,
	id: number,
): Promise<null> {
	return knlgDelete<null>(path(workspaceCode, `/cards/${id}`));
}

export async function publishKnowledgeCard(
	workspaceCode: string,
	id: number,
): Promise<KnowledgeCard> {
	return knlgPost<KnowledgeCard>(path(workspaceCode, `/cards/${id}/publish`));
}

export async function deprecateKnowledgeCard(
	workspaceCode: string,
	id: number,
): Promise<KnowledgeCard> {
	return knlgPost<KnowledgeCard>(path(workspaceCode, `/cards/${id}/deprecate`));
}

export async function listKnowledgeCardVersions(
	workspaceCode: string,
	id: number,
): Promise<KnowledgeCardListResponse> {
	return knlgGet<KnowledgeCardListResponse>(
		path(workspaceCode, `/cards/${id}/versions`),
	);
}

export async function listKnowledgeCardSources(
	workspaceCode: string,
	id: number,
): Promise<SourceRef[]> {
	return knlgGet<SourceRef[]>(path(workspaceCode, `/cards/${id}/sources`));
}

File: neo/backend/src/app/api/v1/knlg_base/knowledge.py
"""Knowledge Base API endpoints."""

from fastapi import APIRouter, Depends, Path
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.models.user import User
from app.schemas.knlg_base import (
    KnowledgeCardCreate,
    KnowledgeCardListResponse,
    KnowledgeCardResponse,
    KnowledgeCardUpdate,
    KnowledgeCardVersionResponse,
    SourceRefResponse,
)
from app.schemas.response import ApiResponse
from app.services.knlg_base.knowledge import KnlgKnowledgeCardService

router = APIRouter(prefix="/knowledge", tags=["knlg-base.knowledge"])


def get_service(db: Session = Depends(get_db)) -> KnlgKnowledgeCardService:
    return KnlgKnowledgeCardService(db)


@router.get("", response_model=ApiResponse[KnowledgeCardListResponse])
def list_cards(
    workspace_code: str = Path(..., description="Workspace code"),
    page: int = 1,
    page_size: int = 20,
    domain: str | None = None,
    type: str | None = None,
    status: str | None = None,
    validation_status: str | None = None,
    keyword: str | None = None,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    """List knowledge cards with filters."""
    items, total = service.list_cards(
        workspace_code,
        current_user,
        page,
        page_size,
        domain=domain,
        type_=type,
        status=status,
        validation_status=validation_status,
        keyword=keyword,
    )
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0
    return ApiResponse.success(
        KnowledgeCardListResponse(
            items=[KnowledgeCardResponse.model_validate(c) for c in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        ),
    )


@router.post("", response_model=ApiResponse[KnowledgeCardResponse])
def create_card(
    workspace_code: str,
    data: KnowledgeCardCreate,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    """Create a new knowledge card."""
    kc = service.create_card(workspace_code, current_user, data.model_dump())
    return ApiResponse.success(KnowledgeCardResponse.model_validate(kc))


@router.get("/cards/{card_id}", response_model=ApiResponse[KnowledgeCardResponse])
def get_card(
    workspace_code: str,
    card_id: int,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    kc = service.get_card(workspace_code, current_user, card_id)
    return ApiResponse.success(KnowledgeCardResponse.model_validate(kc))


@router.put("/cards/{card_id}", response_model=ApiResponse[KnowledgeCardResponse])
def update_card(
    workspace_code: str,
    card_id: int,
    data: KnowledgeCardUpdate,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    kc = service.update_card(workspace_code, current_user, card_id, data.model_dump(exclude_unset=True))
    return ApiResponse.success(KnowledgeCardResponse.model_validate(kc))


@router.delete("/cards/{card_id}")
def delete_card(
    workspace_code: str,
    card_id: int,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    service.delete_card(workspace_code, current_user, card_id)
    return ApiResponse.success(None)


@router.post("/cards/{card_id}/publish", response_model=ApiResponse[KnowledgeCardResponse])
def publish_card(
    workspace_code: str,
    card_id: int,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    kc = service.transition_card(workspace_code, current_user, card_id, "published")
    return ApiResponse.success(KnowledgeCardResponse.model_validate(kc))


@router.post("/cards/{card_id}/deprecate", response_model=ApiResponse[KnowledgeCardResponse])
def deprecate_card(
    workspace_code: str,
    card_id: int,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    kc = service.transition_card(workspace_code, current_user, card_id, "deprecated")
    return ApiResponse.success(KnowledgeCardResponse.model_validate(kc))


@router.get("/cards/{card_id}/versions")
def list_versions(
    workspace_code: str,
    card_id: int,
    page: int = 1,
    page_size: int = 20,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    """List version history (READ-ONLY in P0)."""
    versions = service.list_versions(workspace_code, current_user, card_id)
    # Apply simple pagination in service
    start = (page - 1) * page_size
    end = start + page_size
    items = versions[start:end]
    total = len(versions)
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0
    from app.schemas.knlg_base.knowledge import KnowledgeCardListResponse

    return ApiResponse.success(
        KnowledgeCardListResponse(
            items=[KnowledgeCardVersionResponse.model_validate(v) for v in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        ),
    )


@router.get("/cards/{card_id}/sources", response_model=ApiResponse[list[SourceRefResponse]])
def list_sources(
    workspace_code: str,
    card_id: int,
    service: KnlgKnowledgeCardService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    refs = service.list_source_refs(workspace_code, current_user, card_id)
    return ApiResponse.success([SourceRefResponse.model_validate(r) for r in refs])
